cmd/akamai-purger/main.go

```diff
@@ -215,15 +215,19 @@ func (ap *akamaiPurger) purge() error {
 	return nil
 }
 
-// Purge is an exported gRPC method which receives purge requests and appends
-// them to the queue.
+// Purge is an exported gRPC method which receives purge requests containing
+// URLs and prepends them to the purger queue.
 func (ap *akamaiPurger) Purge(ctx context.Context, req *akamaipb.PurgeRequest) (*emptypb.Empty, error) {
 	ap.Lock()
 	defer ap.Unlock()
-	if len(ap.toPurge) >= ap.maxQueueSize {
-		return nil, errors.New("akamai-purger queue too large")
+	queueLength := len(ap.toPurge)
+	if queueLength >= ap.maxQueueSize {
+		// Drop the oldest entry from the queue to make room for the new
+		// request.
+		ap.toPurge = ap.toPurge[:queueLength-1]
 	}
-	ap.toPurge = append(ap.toPurge, req.Urls)
+	// Add the entry from the new request to the front of the queue.
+	ap.toPurge = append([][]string{req.Urls}, ap.toPurge...)
```

Comment:
There's a subtle but interesting performance problem here! This will always allocate a new backing array and assign it to `ap.toPurge`. If we think about what performance characteristics we'd like, we want the common case to be: we have a backing array with more space than we need, and as we push items onto the stack, we take up additional space in that backing array (increasing a slice's `len`). As we pop elements from the stack, we remove them from the backing array (by decreasing the slice's `len`). A diagram:

```
[a b c _ _ _ _ ]
      ^-len    ^-cap

push(d)

[a b c d _ _ _ ]
        ^-len  ^-cap

pop()

[a b c _ _ _ _ ]
      ^-len    ^-cap
```

In other words, notionally the "top" of the stack should be the right-hand side of the slice, where there is plenty of room to grow.

Also to be clear, I don't think we need to preallocate the backing array to be the max size - we can let it naturally grow to the high-water mark of what we've seen so far.

Reply:
Oh, 🤦🏻‍♀️ this makes a lot more sense.

Reply:
Made a toy to model this and you're right. It results in about 78% less reallocations for growing the backing array.

```go
package main

import (
	"fmt"
)

var (
	frontQueue   []int
	frontPrevCap int
	frontRealloc int
	backQueue    []int
	backPrevCap  int
	backRealloc  int
)

func front(i int) {
	if frontPrevCap != cap(frontQueue) {
		fmt.Printf("%4d: %p FRONT CAP: %d->%d\n", i, frontQueue, frontPrevCap, cap(frontQueue))
		frontPrevCap = cap(frontQueue)
		frontRealloc++
	}
	frontQueue = append([]int{i}, frontQueue...)
}
func back(i int) {
	if backPrevCap != cap(backQueue) {
		fmt.Printf("%4d: %p BACK CAP: %d->%d\n", i, backQueue, backPrevCap, cap(backQueue))
		backPrevCap = cap(backQueue)
		backRealloc++
	}
	backQueue = append(backQueue, i)
}

func main() {

	fmt.Printf("   0:\t%p\tFRONT CAP:\t%d\n", frontQueue, cap(frontQueue))
	frontQueue = append([]int{0}, frontQueue...)

	fmt.Printf("   0:\t%p\tBACK CAP:\t%d\n", backQueue, cap(backQueue))
	backQueue = append(backQueue, 0)

	for i := 1; i < 1250; i++ {
		front(i)
		back(i)
	}

	fmt.Printf("FRONT: %d\tBACK: %d", frontRealloc, backRealloc)
}
```

```text
$ go run ~/Desktop/main.go
   0:   0x0     FRONT CAP:      0
   0:   0x0     BACK CAP:       0
   1: 0x140000aa020 FRONT CAP: 0->1
   1: 0x140000aa028 BACK CAP: 0->1
   2: 0x140000aa040 FRONT CAP: 1->2
   2: 0x140000aa050 BACK CAP: 1->2
   3: 0x140000b0000 FRONT CAP: 2->3
   3: 0x140000ac040 BACK CAP: 2->4
   4: 0x140000ac060 FRONT CAP: 3->4
   5: 0x140000b2000 FRONT CAP: 4->6
   5: 0x140000a2080 BACK CAP: 4->8
   7: 0x140000a20c0 FRONT CAP: 6->8
   9: 0x140000b4000 FRONT CAP: 8->10
   9: 0x140000b6000 BACK CAP: 8->16
  11: 0x1400008c060 FRONT CAP: 10->12
  13: 0x140000b8000 FRONT CAP: 12->14
  15: 0x140000b6080 FRONT CAP: 14->16
  17: 0x140000ba000 FRONT CAP: 16->18
  17: 0x140000bc000 BACK CAP: 16->32
  19: 0x140000be000 FRONT CAP: 18->20
  21: 0x140000c0000 FRONT CAP: 20->22
  23: 0x140000c2000 FRONT CAP: 22->24
  25: 0x140000c4000 FRONT CAP: 24->26
  27: 0x140000c6000 FRONT CAP: 26->28
  29: 0x140000c8000 FRONT CAP: 28->30
  31: 0x140000bc100 FRONT CAP: 30->32
  33: 0x140000ca000 FRONT CAP: 32->36
  33: 0x140000cc000 BACK CAP: 32->64
  37: 0x140000ce000 FRONT CAP: 36->40
  41: 0x140000d0000 FRONT CAP: 40->44
  45: 0x140000d2000 FRONT CAP: 44->48
  49: 0x140000d4000 FRONT CAP: 48->52
  53: 0x140000d6000 FRONT CAP: 52->56
  57: 0x140000d8000 FRONT CAP: 56->60
  61: 0x140000cc200 FRONT CAP: 60->64
  65: 0x140000da000 FRONT CAP: 64->72
  65: 0x140000dc000 BACK CAP: 64->128
  73: 0x140000de000 FRONT CAP: 72->80
  81: 0x140000e0000 FRONT CAP: 80->88
  89: 0x140000e2000 FRONT CAP: 88->96
  97: 0x140000e4000 FRONT CAP: 96->112
 113: 0x140000dc400 FRONT CAP: 112->128
 129: 0x140000ec000 FRONT CAP: 128->144
 129: 0x140000ee000 BACK CAP: 128->256
 145: 0x140000f4000 FRONT CAP: 144->160
 161: 0x140000fa000 FRONT CAP: 160->176
 177: 0x140000fe000 FRONT CAP: 176->192
 193: 0x1400018a000 FRONT CAP: 192->224
 225: 0x140000ee800 FRONT CAP: 224->256
 257: 0x140001aa000 FRONT CAP: 256->288
 257: 0x140001ae000 BACK CAP: 256->512
 289: 0x140001c0000 FRONT CAP: 288->336
 337: 0x140001e0000 FRONT CAP: 336->384
 385: 0x14000206000 FRONT CAP: 384->400
 401: 0x14000216000 FRONT CAP: 400->432
 433: 0x140001af000 FRONT CAP: 432->512
 513: 0x14000282000 FRONT CAP: 512->608
 513: 0x14000288000 BACK CAP: 512->848
 609: 0x14000306000 FRONT CAP: 608->672
 673: 0x1400035e000 FRONT CAP: 672->768
 769: 0x140003ee000 FRONT CAP: 768->816
 817: 0x14000289a80 FRONT CAP: 816->848
 849: 0x14000470000 FRONT CAP: 848->864
 849: 0x1400047c000 BACK CAP: 848->1280
 865: 0x140002fe000 FRONT CAP: 864->1024
1025: 0x14000188000 FRONT CAP: 1024->1184
1185: 0x1400030e000 FRONT CAP: 1184->1216
1217: 0x1400047e800 FRONT CAP: 1216->1280
FRONT: 54       BACK: 12
```